src/psset.c

```diff
@@ -149,15 +178,58 @@ psset_update_begin(psset_t *psset, hpdata_t *ps) {
 	hpdata_assert_consistent(ps);
 	assert(hpdata_in_psset_get(ps));
 	hpdata_updating_set(ps, true);
-	psset_do_alloc_tracking_remove(psset, ps);
+	psset_stats_remove(psset, ps);
+	if (hpdata_in_psset_alloc_container_get(ps)) {
+		/*
+		 * Some metadata updates can break alloc container invariants
+		 * (e.g. the longest free range determines the hpdata_heap_t the
+		 * pageslab lives in).
+		 */
+		assert(hpdata_alloc_allowed_get(ps));
+		psset_alloc_container_remove(psset, ps);
+	}
+	/*
+	 * We don't update presence in the purge list or hugify list; we try to
+	 * keep those FIFO, even in the presence of other metadata updates.
+	 * We'll update presence at the end of the metadata update if necessary.
+	 */
 }
 
 void
 psset_update_end(psset_t *psset, hpdata_t *ps) {
 	hpdata_assert_consistent(ps);
 	assert(hpdata_in_psset_get(ps));
 	hpdata_updating_set(ps, false);
-	psset_do_alloc_tracking_insert(psset, ps);
+	psset_stats_insert(psset, ps);
+
+	/*
+	 * The update begin should have removed ps from whatever alloc container
+	 * it was in.
+	 */
+	assert(!hpdata_in_psset_alloc_container_get(ps));
+	if (hpdata_alloc_allowed_get(ps)) {
```

Comment:
Question: Can you help to clarify the relation between _in_alloc_container and _alloc_allowed (and similarly the other ones below)?

Reply:
Added some comments; the short version is that the latter are used to define what policies are (declaratively) in the HPA, while the former are used to enforce those policies.

For something more concrete: suppose the dalloc causes some hpdata to have new dirty pages; we can just set purge_allowed and the update_end call will take care of the rest; we don't need to worry about whether or not it was previously purgeable to determine the modification to make.